server/src/main/java/io/crate/expression/scalar/ConcatwsFunction.java

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar;
+
+import io.crate.data.Input;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Literal;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public abstract class ConcatwsFunction extends Scalar<String, String> {
+
+    public static final String NAME = "concat_ws";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature()
+            ).withVariableArity(),
+            StringConcatwsFunction::new
+        );
+
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    ConcatwsFunction(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public Symbol normalizeSymbol(Function function, TransactionContext txnCtx, NodeContext nodeCtx) {
+        if (anyNonLiterals(function.arguments())) {
+            return function;
+        }
+        Input[] inputs = new Input[function.arguments().size()];
+        for (int i = 0; i < function.arguments().size(); i++) {
+            inputs[i] = ((Input) function.arguments().get(i));
+        }
+        //noinspection unchecked
+        return Literal.ofUnchecked(boundSignature.getReturnType().createType(), evaluate(txnCtx, nodeCtx, inputs));
+    }
+
+    static class StringConcatwsFunction extends ConcatwsFunction {
+
+        StringConcatwsFunction(Signature signature, Signature boundSignature) {
+            super(signature, boundSignature);
+        }
+
+        @Override
+        public String evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input[] args) {
+
+            String delimitArg = (String) args[0].value();
+            if (delimitArg == null) {
+                throw new NullPointerException("Delimiting argument of the 'concat_ws' function cannot be null.");
+            }
```

Comment:
PostgreSQL returns `null` if the first argument is null. I think we should match the behavior:

```
# select concat_ws(null, 'foo', 'bar');
 concat_ws
-----------
 💩
(1 row)
```

Reply:
"should not be NULL" unfortunately doesn't state how it behaves if it is null. That it fails isn't stated explicitly and the actual behavior is that it returns null (at least in PostgreSQL 13.6, I didn't check with other versions)

Our default policy is to follow PostgreSQL semantics unless there is a good reason to diverge.

Reply:
@mfussenegger Ill make the changes. Thanks for the feedback.